src/bin/fd_readdir.rs

```diff
@@ -66,7 +66,9 @@ fn test_fd_readdir(dir_fd: wasi_unstable::Fd) {
     // Create a file in the scratch directory.
 
     let sl = unsafe { slice::from_raw_parts(buf.as_ptr(), min(BUF_LEN, bufused)) };
-    let mut dirs = ReadDir::from_slice(sl);
+    let mut dirs: Vec<_> = ReadDir::from_slice(sl).collect();
+    dirs.sort_by_key(|d| d.name.clone());
+    let mut dirs = dirs.into_iter();
```

Comment:
Hmm, given that you've managed to prove that depending on the host, `fd_readdir` is not guaranteed to produce entries always in the same order (see [here](https://dev.azure.com/CraneStation/wasi-common/_build/results?buildId=331)), I'm wondering, should we perhaps enforce order that the WASI spec level? That is, have the syscall sort the results before returning them to the embedder? @sunfishcode thoughts?

Reply:
I'm against. Even on newer kernels the results are not sorted, this was a special case of `.` & `..` where it was in a precise order on Linux 4.19. For some use cases the order is completely irrelevant and enforcing order on the spec level imposes an unnecessary sorting cost on the call. If the user needs to have the entries sorted, he should do it himself.

Reply:
Mhm, that's a good point.

Reply:
Relatedly, does anyone know what POSIX means by "ordered" in the first sentence of the description of [`readdir`](http://pubs.opengroup.org/onlinepubs/9699919799/functions/readdir.html)?

I do have some sympathy for taking some overhead in order to help avoid portability problems. However, in the situation here, it might be prohibitively expensive to do this in the case where the user uses a small buffer to read a large directory using multiple `fd_readdir` calls, because we'd have to somehow read the whole directory up front sort it, and then remember that order between `fd_readdir` calls.

Reply:
I guess that it means that the order is implementation-defined but fixed, i.e. the order of directory entries will be the same when the function is executed twice